File: employee-service/src/main/java/com/erp/employee_service/service/promotion/PromotionService.java
package com.erp.employee_service.service.promotion;

import com.erp.employee_service.dto.notification.SendNotificationDto;
import com.erp.employee_service.dto.promotion.PromotionRequestDto;
import com.erp.employee_service.dto.promotion.PromotionResponseDto;
import com.erp.employee_service.dto.promotion.PromotionUpdateDto;
import com.erp.employee_service.entity.department.Department;
import com.erp.employee_service.entity.designation.Designation;
import com.erp.employee_service.entity.Employee;
import com.erp.employee_service.entity.promotion.Promotion;
import com.erp.employee_service.exception.ResourceNotFoundException;
import com.erp.employee_service.repository.DepartmentRepository;
import com.erp.employee_service.repository.DesignationRepository;
import com.erp.employee_service.repository.EmployeeRepository;
import com.erp.employee_service.repository.PromotionRepository;
import com.erp.employee_service.service.notification.NotificationService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class PromotionService {

    private final PromotionRepository promotionRepository;
    private final EmployeeRepository employeeRepository;
    private final DepartmentRepository departmentRepository;
    private final DesignationRepository designationRepository;
    private final NotificationService notificationService;

    public PromotionResponseDto createPromotion(String employeeId, PromotionRequestDto requestDto) {
        // Validate employee
        Employee employee = employeeRepository.findByEmployeeId(employeeId)
                .orElseThrow(() -> new ResourceNotFoundException("Employee not found with ID: " + employeeId));

        // Validate new department
        Department newDepartment = departmentRepository.findById(requestDto.getNewDepartmentId())
                .orElseThrow(() -> new ResourceNotFoundException("Department not found with ID: " + requestDto.getNewDepartmentId()));

        // Validate new designation
        Designation newDesignation = designationRepository.findById(requestDto.getNewDesignationId())
                .orElseThrow(() -> new ResourceNotFoundException("Designation not found with ID: " + requestDto.getNewDesignationId()));

        // Get current department and designation from employee
        Department oldDepartment = employee.getDepartment();
        Designation oldDesignation = employee.getDesignation();

        if (oldDepartment == null || oldDesignation == null) {
            throw new IllegalStateException("Employee does not have current department or designation assigned");
        }

        // Create promotion record
        Promotion promotion = Promotion.builder()
                .employee(employee)
                .oldDepartment(oldDepartment)
                .oldDesignation(oldDesignation)
                .newDepartment(newDepartment)
                .newDesignation(newDesignation)
                .isPromotion(requestDto.getIsPromotion())
                .sendNotification(requestDto.getSendNotification())
                .remarks(requestDto.getRemarks())
                .build();

        Promotion savedPromotion = promotionRepository.save(promotion);

        // Update employee with new department and designation
        employee.setDepartment(newDepartment);
        employee.setDesignation(newDesignation);
        employeeRepository.save(employee);

        // Send notification if requested
        if (requestDto.getSendNotification()) {
            sendPromotionNotification(employee, savedPromotion);
        }

        return mapToResponseDto(savedPromotion);
    }

    public PromotionResponseDto updatePromotion(Long id, PromotionUpdateDto updateDto) {
        // Find existing promotion
        Promotion promotion = promotionRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Promotion record not found with ID: " + id));

        // Validate new department
        Department newDepartment = departmentRepository.findById(updateDto.getNewDepartmentId())
                .orElseThrow(() -> new ResourceNotFoundException("Department not found with ID: " + updateDto.getNewDepartmentId()));

        // Validate new designation
        Designation newDesignation = designationRepository.findById(updateDto.getNewDesignationId())
                .orElseThrow(() -> new ResourceNotFoundException("Designation not found with ID: " + updateDto.getNewDesignationId()));

        Employee employee = promotion.getEmployee();

        // Track old values for notification comparison
        Department oldDepartmentBeforeUpdate = promotion.getNewDepartment();
        Designation oldDesignationBeforeUpdate = promotion.getNewDesignation();
        Boolean oldIsPromotion = promotion.getIsPromotion();

        // Update promotion record
        promotion.setNewDepartment(newDepartment);
        promotion.setNewDesignation(newDesignation);
        promotion.setIsPromotion(updateDto.getIsPromotion());
        promotion.setSendNotification(updateDto.getSendNotification());
        promotion.setRemarks(updateDto.getRemarks());

        Promotion updatedPromotion = promotionRepository.save(promotion);

        // Update employee with new department and designation
        employee.setDepartment(newDepartment);
        employee.setDesignation(newDesignation);
        employeeRepository.save(employee);


        return mapToResponseDto(updatedPromotion);
    }


    @Async
    public CompletableFuture<Void> sendPromotionNotification(Employee employee, Promotion promotion) {
        try {
            SendNotificationDto dto = new SendNotificationDto();
            dto.setReceiverEmployeeId(employee.getEmployeeId());

            if (promotion.getIsPromotion()) {
                dto.setTitle("Promotion Notification");
                dto.setMessage("Congratulations! You have been promoted from " +
                        promotion.getOldDesignation().getDesignationName() + " (" + promotion.getOldDepartment().getDepartmentName() +
                        ") to " + promotion.getNewDesignation().getDesignationName() + " (" +
                        promotion.getNewDepartment().getDepartmentName() + ").");
            } else {
                dto.setTitle("Demotion Notification");
                dto.setMessage("Your designation has been changed from " +
                        promotion.getOldDesignation().getDesignationName() + " (" + promotion.getOldDepartment().getDepartmentName() +
                        ") to " + promotion.getNewDesignation().getDesignationName() + " (" +
                        promotion.getNewDepartment().getDepartmentName() + ").");
            }

            dto.setType("PROMOTION");
            notificationService.sendNotification(null, dto);
        } catch (Exception e) {
            System.err.println("Failed to send promotion notification: " + e.getMessage());
        }
        return CompletableFuture.completedFuture(null);
    }

    public List<PromotionResponseDto> getAllPromotions() {
        return promotionRepository.findAllByOrderByCreatedAtDesc()
                .stream()
                .map(this::mapToResponseDto)
                .collect(Collectors.toList());
    }

    public List<PromotionResponseDto> getPromotionsByEmployee(String employeeId) {
        return promotionRepository.findByEmployee_EmployeeId(employeeId)
                .stream()
                .map(this::mapToResponseDto)
                .collect(Collectors.toList());
    }

    public PromotionResponseDto getPromotionById(Long id) {
        Promotion promotion = promotionRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Promotion record not found with ID: " + id));
        return mapToResponseDto(promotion);
    }

    public void deletePromotion(Long id) {
        Promotion promotion = promotionRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Promotion record not found with ID: " + id));
        promotionRepository.delete(promotion);
    }

    private PromotionResponseDto mapToResponseDto(Promotion promotion) {
        PromotionResponseDto dto = new PromotionResponseDto();
        dto.setId(promotion.getId());
        dto.setEmployeeId(promotion.getEmployee().getEmployeeId());
        dto.setEmployeeName(promotion.getEmployee().getName());
        dto.setOldDepartmentId(promotion.getOldDepartment().getId());
        dto.setOldDepartmentName(promotion.getOldDepartment().getDepartmentName());
        dto.setOldDesignationId(promotion.getOldDesignation().getId());
        dto.setOldDesignationName(promotion.getOldDesignation().getDesignationName());
        dto.setNewDepartmentId(promotion.getNewDepartment().getId());
        dto.setNewDepartmentName(promotion.getNewDepartment().getDepartmentName());
        dto.setNewDesignationId(promotion.getNewDesignation().getId());
        dto.setNewDesignationName(promotion.getNewDesignation().getDesignationName());
        dto.setIsPromotion(promotion.getIsPromotion());
        dto.setSendNotification(promotion.getSendNotification());
        dto.setCreatedAt(promotion.getCreatedAt());
        dto.setRemarks(promotion.getRemarks());
        return dto;
    }

}

File: erp-node-backend/services/employee-service/src/services/promotion.service.ts
import type { Department, Designation, Employee, Promotion, PrismaClient } from "@prisma/client";

import { HttpError } from "../common/errors.js";
import { NotificationService } from "./notification.service.js";

type PromotionWithRelations = Promotion & {
  employee: Employee;
  oldDepartment: Department;
  oldDesignation: Designation;
  newDepartment: Department;
  newDesignation: Designation;
};

const promotionInclude = {
  employee: true,
  oldDepartment: true,
  oldDesignation: true,
  newDepartment: true,
  newDesignation: true
} as const;

export class PromotionService {
  constructor(
    private readonly prisma: PrismaClient,
    private readonly notificationService: NotificationService
  ) {}

  async createPromotion(
    employeeId: string,
    request: {
      newDepartmentId?: number;
      newDesignationId?: number;
      sendNotification?: boolean;
      isPromotion?: boolean;
      remarks?: string | null;
    },
    actorEmployeeId: string
  ): Promise<Record<string, unknown>> {
    const employee = await this.requireEmployee(employeeId);
    const sendNotification = request.sendNotification;
    const isPromotion = request.isPromotion;

    if (!employee.departmentId || !employee.designationId) {
      throw new HttpError(400, "Employee does not have current department or designation assigned");
    }

    if (!request.newDepartmentId || !request.newDesignationId || sendNotification === undefined || isPromotion === undefined) {
      throw new HttpError(400, "newDepartmentId, newDesignationId, sendNotification and isPromotion are required");
    }

    const [newDepartment, newDesignation] = await Promise.all([
      this.requireDepartment(BigInt(request.newDepartmentId)),
      this.requireDesignation(BigInt(request.newDesignationId))
    ]);

    const promotionId = await this.prisma.$transaction(async (tx) => {
      const saved = await tx.promotion.create({
        data: {
          employeeId: employee.employeeId,
          oldDepartmentId: employee.departmentId!,
          oldDesignationId: employee.designationId!,
          newDepartmentId: newDepartment.id,
          newDesignationId: newDesignation.id,
          isPromotion,
          sendNotification,
          remarks: request.remarks?.trim() || null
        }
      });

      await tx.employee.update({
        where: { employeeId: employee.employeeId },
        data: {
          departmentId: newDepartment.id,
          designationId: newDesignation.id
        }
      });

      return saved.id;
    });

    const promotion = await this.getPromotionRecord(Number(promotionId));

    if (sendNotification) {
      await this.notificationService.sendNotification(actorEmployeeId, {
        receiverEmployeeId: employee.employeeId,
        title: isPromotion ? "Promotion Notification" : "Demotion Notification",
        message: isPromotion
          ? `Congratulations! You have been promoted from ${promotion.oldDesignation.designationName} (${promotion.oldDepartment.departmentName}) to ${promotion.newDesignation.designationName} (${promotion.newDepartment.departmentName}).`
          : `Your designation has been changed from ${promotion.oldDesignation.designationName} (${promotion.oldDepartment.departmentName}) to ${promotion.newDesignation.designationName} (${promotion.newDepartment.departmentName}).`,
        type: "PROMOTION"
      });
    }

    return mapPromotion(promotion);
  }

  async updatePromotion(
    id: number,
    request: {
      newDepartmentId?: number;
      newDesignationId?: number;
      sendNotification?: boolean;
      isPromotion?: boolean;
      remarks?: string | null;
    },
    _actorEmployeeId: string
  ): Promise<Record<string, unknown>> {
    const existing = await this.getPromotionRecord(id);
    const sendNotification = request.sendNotification;
    const isPromotion = request.isPromotion;

    if (!request.newDepartmentId || !request.newDesignationId || sendNotification === undefined || isPromotion === undefined) {
      throw new HttpError(400, "newDepartmentId, newDesignationId, sendNotification and isPromotion are required");
    }

    const [newDepartment, newDesignation] = await Promise.all([
      this.requireDepartment(BigInt(request.newDepartmentId)),
      this.requireDesignation(BigInt(request.newDesignationId))
    ]);

    const updated = await this.prisma.$transaction(async (tx) => {
      const saved = await tx.promotion.update({
        where: { id: BigInt(id) },
        data: {
          newDepartmentId: newDepartment.id,
          newDesignationId: newDesignation.id,
          sendNotification,
          isPromotion,
          remarks: request.remarks?.trim() || null
        },
        include: promotionInclude
      });

      await tx.employee.update({
        where: { employeeId: existing.employeeId },
        data: {
          departmentId: newDepartment.id,
          designationId: newDesignation.id
        }
      });

      return saved;
    });

    return mapPromotion(updated);
  }

  async getAllPromotions(): Promise<Record<string, unknown>[]> {
    const promotions = await this.prisma.promotion.findMany({
      include: promotionInclude,
      orderBy: { createdAt: "desc" }
    });

    return promotions.map(mapPromotion);
  }

  async getPromotionsByEmployee(employeeId: string): Promise<Record<string, unknown>[]> {
    const promotions = await this.prisma.promotion.findMany({
      where: { employeeId: employeeId.trim().toUpperCase() },
      include: promotionInclude,
      orderBy: { createdAt: "desc" }
    });

    return promotions.map(mapPromotion);
  }

  async getPromotionById(id: number): Promise<Record<string, unknown>> {
    return mapPromotion(await this.getPromotionRecord(id));
  }

  async deletePromotion(id: number): Promise<void> {
    await this.getPromotionRecord(id);
    await this.prisma.promotion.delete({
      where: { id: BigInt(id) }
    });
  }

  private async requireEmployee(employeeId: string): Promise<Employee> {
    const employee = await this.prisma.employee.findUnique({
      where: { employeeId: employeeId.trim().toUpperCase() }
    });

    if (!employee) {
      throw new HttpError(404, `Employee not found with ID: ${employeeId}`);
    }

    return employee;
  }

  private async requireDepartment(id: bigint): Promise<Department> {
    const department = await this.prisma.department.findUnique({
      where: { id }
    });

    if (!department) {
      throw new HttpError(404, `Department not found with ID: ${id}`);
    }

    return department;
  }

  private async requireDesignation(id: bigint): Promise<Designation> {
    const designation = await this.prisma.designation.findUnique({
      where: { id }
    });

    if (!designation) {
      throw new HttpError(404, `Designation not found with ID: ${id}`);
    }

    return designation;
  }

  private async getPromotionRecord(id: number): Promise<PromotionWithRelations> {
    const promotion = await this.prisma.promotion.findUnique({
      where: { id: BigInt(id) },
      include: promotionInclude
    });

    if (!promotion) {
      throw new HttpError(404, `Promotion record not found with ID: ${id}`);
    }

    return promotion;
  }
}

function mapPromotion(promotion: PromotionWithRelations): Record<string, unknown> {
  return {
    id: Number(promotion.id),
    employeeId: promotion.employee.employeeId,
    employeeName: promotion.employee.name,
    oldDepartmentId: Number(promotion.oldDepartmentId),
    oldDepartmentName: promotion.oldDepartment.departmentName,
    oldDesignationId: Number(promotion.oldDesignationId),
    oldDesignationName: promotion.oldDesignation.designationName,
    newDepartmentId: Number(promotion.newDepartmentId),
    newDepartmentName: promotion.newDepartment.departmentName,
    newDesignationId: Number(promotion.newDesignationId),
    newDesignationName: promotion.newDesignation.designationName,
    isPromotion: promotion.isPromotion,
    sendNotification: promotion.sendNotification,
    createdAt: promotion.createdAt,
    remarks: promotion.remarks
  };
}
